_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+  Terraform) that are optimized for Terraform workflows. Both services provide a web UI for remotely running `terraform
+  plan` and `terraform apply`, and can be integrated with VCS platforms to implement CI/CD workflows that remotely run
+  `terraform` in reaction to git. Additionally, Terraform Enterprise in particular provides the ability to manage
+  deployment policies and permissions as code. Note that Terraform Cloud and Enterprise are both optimized for running
+  `terraform`. That is, they do not support the use of external binaries such as `terragrunt` or `docker`, even when
+  called from within `terraform` using a `local-exec` provisioner or external data source. In terms of the threat model,
+  Terraform Cloud and Enterprise both implement respectable security practices, including end to end encryption backed
+  by HashiCorp Vault, and you have the option to run in your own data center with Terraform Enterprise to lock it down.
+  Note that Terraform Enterprise relies on API based integrations with the VCS platforms instead of webhooks, so it is
+  not necessary to have a publicly facing service when running in self-hosted mode.
+
+
+[[production_grade_design]]
+== Production-grade design
+
+With all the core concepts out of the way, let's now discuss how to configure a production-grade CI/CD workflow for
+infrastructure code, using a platform that looks something like this:
+
+.Architecture of platform for running Terraform/Terragrunt CI/CD workflows.
+image::/assets/img/guides/infrastructure-cicd-pipeline/tftg-pipeline-architecture.png[]
+
+.Sequence diagram of running Terraform/Terragrunt CI/CD workflows.
+image::/assets/img/guides/infrastructure-cicd-pipeline/tftg-pipeline-sequence-diagram.png[]
+
+
+[[use_generic_cicd_platforms_as_a_workflow_engine_but_run_infrastructure_deployments_from_within_your_account]]
+=== Use generic CI/CD platforms as a workflow engine but run infrastructure deployments from within your account
+
+As we covered in the <<cicd_platforms>> section of the Core Concepts, all CI/CD platforms support some form of
+integration with VCS repositories. Additionally, CI/CD platforms are optimized for implementing workflows and pipelines
+that react to VCS activities. They all support a way to start executing a pipeline on commits, approval steps in the
+pipeline, and a way to report back to the VCS repository by annotating commits with status of a pipeline execution.
+With the abundance of CI/CD platforms out there, you are bound to find one that suits your needs. It doesn't matter
+which one you use, but be sure to choose one that has enough support for your application code so that you can use the
+same platform for both application and infrastructure CI/CD workflows. In addition, we recommend selecting a Saas
+based CI/CD platform to avoid the overhead of managing additional infrastructure for the VCS integration and workflow
```

Comment:
I think I'd frame these two paragraphs slightly differently:

1. Use any generic CI / CD server you wish.
1. But don't give that CI / CD server (admin) permissions to deploy arbitrary infra. CI / CD servers aren't secure enough, and you don't want a server that is used for executing arbitrary code and regularly used (and written to) by your entire dev team to have admin permissions. 
1. Instead, give permissions to something within your AWS account that cannot be accessed by anyone but a few admins plus a limited trigger we'll define below. Our concrete implementation is a Fargate-based Task that can only be triggered in limited ways by a Lambda function. Note: I'm not sure even TFE fits this description as you typically give your whole team access to it.
1. You then trigger your locked-down system from the CI server and stream the logs. If anyone gets access to your CI server, they can at most kick off builds on existing code, but they don't get arbitrary admin access.

Reply:
Revamped: 427a56bc1a9eb400331fdb34cf82e44a9d6aabcb